service/history/replication/executable_sync_hsm_task.go

```diff
@@ -0,0 +1,241 @@
+// we may want to do some refactoring later.
+
+type (
+	ExecutableSyncHSMTask struct {
+		ProcessToolBox
+
+		definition.WorkflowKey
+		ExecutableTask
+
+		taskAttr *replicationspb.SyncHSMAttributes
+
+		markPoisonPillAttempts int
+	}
+)
+
+var _ ctasks.Task = (*ExecutableSyncHSMTask)(nil)
+var _ TrackableExecutableTask = (*ExecutableSyncHSMTask)(nil)
+
+// var _ BatchableTask = (*ExecutableSyncHSMTask)(nil)
+
+func NewExecutableSyncHSMTask(
+	processToolBox ProcessToolBox,
+	taskID int64,
+	taskCreationTime time.Time,
+	task *replicationspb.SyncHSMAttributes,
+	sourceClusterName string,
+	priority enumsspb.TaskPriority,
+) *ExecutableSyncHSMTask {
+	return &ExecutableSyncHSMTask{
+		ProcessToolBox: processToolBox,
+
+		WorkflowKey: definition.NewWorkflowKey(task.NamespaceId, task.WorkflowId, task.RunId),
+		ExecutableTask: NewExecutableTask(
+			processToolBox,
+			taskID,
+			metrics.SyncHSMTaskScope,
+			taskCreationTime,
+			time.Now().UTC(),
+			sourceClusterName,
+			priority,
+		),
+		taskAttr:               task,
+		markPoisonPillAttempts: 0,
+	}
+}
+
+func (e *ExecutableSyncHSMTask) QueueID() interface{} {
+	return e.WorkflowKey
+}
+
+func (e *ExecutableSyncHSMTask) Execute() error {
+	if e.TerminalState() {
+		return nil
+	}
+
+	namespaceName, apply, nsError := e.GetNamespaceInfo(headers.SetCallerInfo(
+		context.Background(),
+		headers.SystemPreemptableCallerInfo,
+	), e.NamespaceID)
+	if nsError != nil {
+		return nsError
+	} else if !apply {
+		e.Logger.Warn("Skipping the replication task",
+			tag.WorkflowNamespaceID(e.NamespaceID),
+			tag.WorkflowID(e.WorkflowID),
+			tag.WorkflowRunID(e.RunID),
+			tag.TaskID(e.ExecutableTask.TaskID()),
+		)
+		metrics.ReplicationTasksSkipped.With(e.MetricsHandler).Record(
+			1,
+			metrics.OperationTag(metrics.SyncHSMTaskScope),
+			metrics.NamespaceTag(namespaceName),
+		)
+		return nil
+	}
+	ctx, cancel := newTaskContext(namespaceName, e.Config.ReplicationTaskApplyTimeout())
+	defer cancel()
+
+	shardContext, err := e.ShardController.GetShardByNamespaceWorkflow(
+		namespace.ID(e.NamespaceID),
+		e.WorkflowID,
+	)
+	if err != nil {
+		return err
+	}
+	engine, err := shardContext.GetEngine(ctx)
+	if err != nil {
+		return err
+	}
+	return engine.SyncHSM(ctx, &shard.SyncHSMRequest{
+		WorkflowKey:         e.WorkflowKey,
+		StateMachineNode:    e.taskAttr.StateMachineNode,
+		EventVersionHistory: e.taskAttr.VersionHistory,
+	})
+
+}
+
+func (e *ExecutableSyncHSMTask) HandleErr(err error) error {
+	switch retryErr := err.(type) {
+	case nil, *serviceerror.NotFound:
+		return nil
+	case *serviceerrors.RetryReplication:
+		namespaceName, _, nsError := e.GetNamespaceInfo(headers.SetCallerInfo(
+			context.Background(),
+			headers.SystemPreemptableCallerInfo,
+		), e.NamespaceID)
+		if nsError != nil {
+			return err
+		}
+		ctx, cancel := newTaskContext(namespaceName, e.Config.ReplicationTaskApplyTimeout())
+		defer cancel()
+
+		if doContinue, resendErr := e.Resend(
```

Comment:
do you need to update resend to resend state?

Reply:
In full state-based replication world, there will no resend, just state sync (with events backfilling).
For now, no because the task itself is a sync state task. and as long as no task go to DLQ we are good.